### PR TITLE
Multi image support for GRPO/RLOO

### PR DESCRIPTION
**Pull request by qgallouedec**
This PR is the second of a sequence of PR (after #4111) that aims to refactor the generation part of GRPO/RLOO to allow for easier customization. 

While refactoring, I realized that having a clean multi-image support help having a cleaner separation between functions.

try with 

```python
from datasets import load_dataset

from trl import GRPOConfig, GRPOTrainer


# If not handled properly, prompt truncation may truncate image token
dataset = load_dataset("trl-internal-testing/zen-multi-image", "conversational_prompt_only", split="train")

dataset = dataset.filter(lambda x: len(x["images"]) > 0) # currently, mixing samples with and without images is not supported

def my_reward_function(prompts, completions, **kwargs):
    return [1.0] * len(prompts)

training_args = GRPOConfig(
    output_dir="tmp_dir",   
    learning_rate=0.1,  # increase the learning rate to speed up the test
    per_device_train_batch_size=6,  # reduce the batch size to reduce memory usage
    num_generations=3,  # reduce the number of generations to reduce memory usage
    max_completion_length=8,  # reduce the completion length to reduce memory usage
    max_prompt_length=32,
    report_to="none",
)
trainer = GRPOTrainer(
    model="Qwen/Qwen2-VL-2B-Instruct",
    reward_funcs=my_reward_function,  # define a dummy reward function
    args=training_args,
    train_dataset=dataset,
)

previous_trainable_params = {n: param.clone() for n, param in trainer.model.named_parameters()}

trainer.train()
```

### CODE REVIEW

File: tests/test_grpo_trainer.py [hunk 1]
```diff
@@ -1269,7 +1273,7 @@ def test_training_vlm(self, model_id):
         )
         trainer = GRPOTrainer(
             model=model_id,
-            reward_funcs="trl-internal-testing/tiny-Qwen2ForSequenceClassification-2.5",
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py:1272]: we don't support visual reward model, so it doesn't really make sense to test this case, where the image is dropped and a warning is raised.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1020,6 +1019,14 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
...
                 if isinstance(reward_func, nn.Module):  # Module (no PretrainedModel) for compat with compiled models
                     if is_conversational(inputs[0]):
+                        # VLM reward models aren't supported yet, so we drop the image and raise a warning if needed
+                        for prompt in prompts:
+                            for turn in prompt:
+                                if isinstance(turn["content"], list):
+                                    logger.warning_once("Visual reward models aren't supported yet; dropping image.")
+                                    turn["content"] = " ".join(
+                                        e["text"] for e in turn["content"] if e["type"] == "text"
+                                    )
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: from 

```python
[{"role": "user", "content": [{"type": "image"}, {"type": "text", "text": "What color is the sky?"}]}]
```

to 
```python
[{"role": "user", "content": "What color is the sky?"}]
```

plus raise warning

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1065,22 +1072,22 @@ def _generate_and_score_completions(
 
         prompts = [x["prompt"] for x in inputs]
 
-        # We don't yet support visual reward models/function, so we keep a copy of the original text-only prompts for
-        # later use in the reward computation. If images are present, we insert {"type": "image"} as required by the
-        # VLM chat template.
-        original_prompts = copy.deepcopy(prompts)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1071]: instead of keeping the original prompt, we just drop the image later, and raise a warning, see https://github.com/huggingface/trl/pull/4113/files#r2364899902

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1433,7 +1438,7 @@ def _generate_and_score_completions(
         # Calculate rewards for each reward function. rewards_per_func aggregates rewards across all processes. This is
         # important because rewards will be normalized per group, and completions are distributed. We will later slice
         # rewards_per_func to extract each process's subset.
-        rewards_per_func = self._calculate_rewards(inputs, original_prompts, completions, completion_ids_list)
+        rewards_per_func = self._calculate_rewards(inputs, prompts, completions, completion_ids_list)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1443]: see https://github.com/huggingface/trl/pull/4113/files#r2364900545

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -1790,14 +1797,11 @@ def log(self, logs: dict[str, float], start_time: Optional[float] = None) -> Non
...
-                        if img is not None:
-                            # Convert images to wandb Image objects for proper visualization
-                            table["image"].append(wandb.Image(img))
-                        else:
-                            table["image"].append(None)
+                if self._logs["images"]:
+                    table["images"] = []
+                    for image_list in self._logs["images"]:
+                        # Convert images to wandb Image objects for proper visualization
+                        table["images"].append([wandb.Image(image) for image in image_list])
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1807]: tested

File: trl/trainer/utils.py [hunk 6]
```diff
@@ -1780,20 +1781,23 @@ def identity(x):
...
     if sum(lengths) != pixel_values.size(0):
         raise ValueError(f"Mismatch: sum(lengths) = {sum(lengths)} != pixel_values.size(0) = {pixel_values.size(0)}")
 
-    split_values = list(torch.split(batch["pixel_values"], lengths, dim=0))
-    return {**batch, "pixel_values": split_values}
+    boundaries = [0, *accumulate(batch["num_images"])]  # [3, 4, 5] -> [0, 3, 7, 12]
+    sections = [sum(lengths[boundaries[i] : boundaries[i + 1]]) for i in range(len(batch["num_images"]))]
+    split_values = list(torch.split(batch["pixel_values"], sections, dim=0))
+    image_grid_thw = list(torch.split(batch["image_grid_thw"], batch["num_images"], dim=0))
+    return {**batch, "pixel_values": split_values, "image_grid_thw": image_grid_thw}
```
Comment (qgallouedec) [inline trl/trainer/utils.py:1800]: instead of keeping `image_grid_thw` as is, we need to split it depending on the number of images. It gets concatenated later in `_get_per_token_logps_and_entropies` (see line 807)

File: trl/trainer/grpo_trainer.py [hunk 7]
```diff
@@ -804,9 +803,9 @@ def _get_per_token_logps_and_entropies(
             model_inputs = {"input_ids": input_ids_batch, "attention_mask": attention_mask_batch}
 
             if image_grid_thw is not None and pixel_values is not None:
-                model_inputs["image_grid_thw"] = image_grid_thw[start : start + batch_size]
-                start_pixel_idx = image_grid_thw[:start].prod(-1).sum().item()
-                end_pixel_idx = image_grid_thw[: start + batch_size].prod(-1).sum().item()
+                model_inputs["image_grid_thw"] = torch.cat(image_grid_thw[start : start + batch_size])
+                start_pixel_idx = 0 if start == 0 else torch.cat(image_grid_thw[:start]).prod(-1).sum().item()
+                end_pixel_idx = torch.cat(image_grid_thw[: start + batch_size]).prod(-1).sum().item()
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: See https://github.com/huggingface/trl/pull/4113/files#r2364904060, `image_grid_thw` is not a tensor anymore, but a list of tensor

File: trl/trainer/grpo_trainer.py (hunk 5 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:1807]: Screenshot 2025-09-19 at 5 30 14 PM

File: tests/test_grpo_trainer.py [hunk 8]
```diff
@@ -1519,7 +1584,20 @@ def test_training_vlm_and_vllm(self, model_id) -> None:
 
         self.assertIsNotNone(trainer.state.log_history[-1]["train_loss"])
 
+        # Check that the params have changed
+        # Because of the way the tiny models are initialized, the gradient does not flow properly through the
+        # vision parts of the model, so we skip them. Ideally, we should fix the init of these models.
+        params_to_skip = (
+            # "model.vision_tower.",
```
Comment (lewtun) [inline tests/test_grpo_trainer.py]: These are commented out - restore?

File: tests/test_rloo_trainer.py [hunk 9]
```diff
@@ -1265,7 +1285,44 @@ def test_training_vlm_and_vllm(self, model_id) -> None:
...
+            train_dataset=dataset,
+        )
+
+        previous_trainable_params = {n: param.clone() for n, param in trainer.model.named_parameters()}
+
+        trainer.train()
+
+        self.assertIsNotNone(trainer.state.log_history[-1]["train_loss"])
+
+        for n, param in previous_trainable_params.items():
```
Comment (lewtun) [inline tests/test_rloo_trainer.py:1297]: Does the same comment for GRPO apply here? https://github.com/huggingface/trl/pull/4113/files#diff-96dca172e696190fc3e1469166e88aface95ebae959284c6806f2e25d2217c16R1587

File: trl/trainer/grpo_trainer.py [hunk 10]
```diff
@@ -1020,6 +1024,14 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
             with profiling_context(self, reward_func_name):
                 if isinstance(reward_func, nn.Module):  # Module (no PretrainedModel) for compat with compiled models
                     if is_conversational(inputs[0]):
+                        # VLM reward models aren't supported yet, so we drop the image and raise a warning if needed
+                        for prompt in prompts:
+                            for turn in prompt:
+                                if isinstance(turn["content"], list):
+                                    logger.warning_once("Visual reward models aren't supported yet; dropping image.")
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py]: Would raising an error be better than a warning? Otherwise I could imagine the warning could be missed and the training "fails silently" because the reward is only computed on the text part.

File: trl/trainer/rloo_trainer.py [hunk 11]
```diff
@@ -1588,14 +1607,11 @@ def log(self, logs: dict[str, float], start_time: Optional[float] = None) -> Non
...
-                        if img is not None:
-                            # Convert images to wandb Image objects for proper visualization
-                            table["image"].append(wandb.Image(img))
-                        else:
-                            table["image"].append(None)
+                if self._logs["images"]:
+                    table["images"] = []
+                    for image_list in self._logs["images"]:
+                        # Convert images to wandb Image objects for proper visualization
+                        table["images"].append([wandb.Image(image) for image in image_list])
```
Comment (lewtun) [inline trl/trainer/rloo_trainer.py:1606]: At some point it would be nice to also add the `trackio` variant for table images

File: tests/test_grpo_trainer.py (hunk 8 above)
Comment (qgallouedec) [reply tests/test_grpo_trainer.py]: ah yes, actaully this can be removed, the test seems to pass

File: tests/test_rloo_trainer.py (hunk 9 above)
Comment (qgallouedec) [reply tests/test_rloo_trainer.py:1297]: answered here https://github.com/huggingface/trl/pull/4113#discussion_r2370659373

File: trl/trainer/grpo_trainer.py (hunk 10 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: Yes, I see. I wonder if anyone would want to train a VLM with a standard LM reward model (ie, not visual reward model). But so far, I've never seen that. We could always support it in the future if there is demand for it. I'll remove this warning, and if the user tries it, the rendering of the chat template will fail, which will prevent from ending in the case of the training failing silently that you describe.
